src/Enum.php

```diff
@@ -2,425 +2,141 @@
-        }
-
-        $this->name = $name;
-        $this->value = $value;
-        $this->index = $index;
-    }
-
-    public function __call(string $name, array $arguments)
+    public static function toArray(): array
     {
-        if (static::startsWith($name, 'is')) {
-            return $this->isEqual(substr($name, 2));
-        }
+        $array = [];
 
-        if (static::isValidName($name)) {
-            return static::make($name);
+        foreach (static::resolveDefinition() as $definition) {
+            $array[$definition->value] = $definition->label;
         }
 
-        throw new BadMethodCallException('Call to undefined method '.static::class.'->'.$name.'()');
+        return $array;
     }
 
     /**
-     * @param string $name
-     * @param mixed[] $arguments
-     *
-     * @return \Spatie\Enum\Enumerable|bool
+     * @param string|int $value
      */
-    public static function __callStatic(string $name, array $arguments)
+    public function __construct($value)
     {
-        if (static::startsWith($name, 'is')) {
-            if (! isset($arguments[0])) {
-                throw new ArgumentCountError('Calling '.static::class.'::'.$name.'() in static context requires one argument');
-            }
+        $definition = $this->findDefinition($value);
 
-            return static::make($arguments[0])->$name();
-        }
+        if ($definition === null) {
+            $enumClass = static::class;
 
-        if (static::isValidName($name) || static::isValidValue($name)) {
-            return static::make($name);
+            throw new BadMethodCallException("There's no value {$value} defined for enum {$enumClass}, consider adding it in the docblock definition.");
         }
 
-        throw new BadMethodCallException('Call to undefined method '.static::class.'::'.$name.'()');
+        $this->value = $definition->value;
+        $this->label = $definition->label;
     }
 
-    public function __toString(): string
+    public function __get($name)
     {
-        return $this->getValue();
-    }
-
-    public function getIndex(): int
-    {
-        return $this->index;
-    }
-
-    public static function getIndices(): array
-    {
-        return array_column(static::resolve(), 'index');
-    }
+        if ($name === 'label') {
+            return $this->label;
+        }
 
-    public function getValue(): string
-    {
-        return $this->value;
+        if ($name === 'value') {
+            return $this->value;
+        }
     }
 
-    public static function getValues(): array
+    public static function __callStatic(string $name, array $arguments)
     {
-        return array_column(static::resolve(), 'value');
+        return new static($name);
     }
 
-    public function getName(): string
+    public function __call($name, $arguments)
     {
-        return $this->name;
-    }
+        if (strpos($name, 'is') === 0) {
+            $other = new static(lcfirst(str_replace('is', '', $name)));
```

Comment:
The `lcfirst` isn't needed because methods can start with uppercase and we also use `strtolower` before instantiating one.

Reply:
you're right!

Reply:
https://github.com/spatie/enum/pull/56/commits/804e224c27ae96c2d843033603ade18329a86ef9